robot-server/robot_server/subsystems/router.py

```diff
@@ -0,0 +1,358 @@
+    responses={status.HTTP_200_OK: {"model": SimpleMultiBody[UpdateProgressSummary]}},
+)
+async def get_subsystem_updates(
+    update_manager: FirmwareUpdateManager = Depends(get_firmware_update_manager),
+) -> PydanticResponse[SimpleMultiBody[UpdateProgressSummary]]:
+    """Return all currently-running firmware update process summaries."""
+    handles = await update_manager.all_ongoing_processes()
+    data = [
+        UpdateProgressSummary.construct(
+            id=handle.process_details.update_id,
+            subsystem=handle.process_details.subsystem,
+            updateStatus=handle.cached_state,
+            createdAt=handle.process_details.created_at,
+        )
+        for handle in handles
+    ]
+    meta = MultiBodyMeta(cursor=0, totalLength=len(data))
+    return await PydanticResponse.create(
+        content=SimpleMultiBody.construct(data=data, meta=meta)
+    )
+
+
+@subsystems_router.get(
+    "/subsystems/updates/current/{subsystem}",
+    summary="Get any currently-ongoing update for a specific subsystem.",
+    description="As /subsystems/updates/current but filtered by the route parameter.",
+    responses={
+        status.HTTP_200_OK: {"model": SimpleBody[UpdateProgressData]},
+        status.HTTP_404_NOT_FOUND: {"model": ErrorBody[NoOngoingUpdate]},
+    },
+)
+async def get_subsystem_update(
+    subsystem: SubSystem,
+    update_manager: FirmwareUpdateManager = Depends(get_firmware_update_manager),
+) -> PydanticResponse[SimpleBody[UpdateProgressData]]:
+    """Return full data about a specific currently-running update process."""
+    try:
+        handle = await update_manager.get_ongoing_update_process_handle_by_subsystem(
+            subsystem
+        )
+    except _NoOngoingUpdate as e:
+        raise NoOngoingUpdate(detail=subsystem.value).as_error(
+            status.HTTP_404_NOT_FOUND
+        ) from e
+    progress = await handle.get_progress()
+    return await PydanticResponse.create(
+        content=SimpleBody.construct(
+            data=UpdateProgressData.construct(
+                id=handle.process_details.update_id,
+                createdAt=handle.process_details.created_at,
+                subsystem=handle.process_details.subsystem,
+                updateStatus=progress.state,
+                updateProgress=progress.progress,
+                updateError=_error_str(progress.error),
+            )
+        )
+    )
+
+
+@subsystems_router.get(
+    "/subsystems/updates/all",
+    summary="Get a list of all updates by id.",
+    description="Get a list of all updates, including both current and concluded ones, by their update id. This is a list of all previous updates since the machine booted, including their final status and whether they succeeded or failed. While an update might complete while you're not polling and therefore disappear from /subsystems/updates/running, you can always pull a previous update by id from here.",
```

Comment:
A couple docstring suggestions:

"Get a list of all updates by ID" means, to me, it will return an object whose keys are the update IDs. But it looks like this is actually returning a list (which is good).

Also, I don't see a `/running` endpoint, but I think it refers to `/current`.

```suggestion
    summary="Get a list of all updates",
    description="Get a list of all updates, including both current and concluded ones, since the machine booted. This includes their final status and whether they succeeded or failed. While an update might complete while you're not polling and therefore disappear from /subsystems/updates/current, you can always pull a previous update by id from here.",
```

Reply:
re: the ids - I had not considered the routes with multiple ids. i think maybe subclassing is a good solution there? i think we want to keep machine readability, but i will also say we could put "which thing is wrong" somewhere other than the title of the exception in a machine-readable way.